File: core/src/main/java/net/wiredtomato/waygl/core/IconInjector.java
package net.wiredtomato.waygl.core;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.imageio.ImageIO;
import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;


/**
 * From [moehreag/wayland_fixes](https://github.com/moehreag/wayland-fixes) DesktopFileInjector
 */
public final class IconInjector {
    private static final Logger LOGGER = LoggerFactory.getLogger("WayGL/IconInjector");

    private static final IconInjector INSTANCE = new IconInjector();

    public static final String APP_ID = "com.mojang.minecraft";

    private static final String ICON_FILE_NAME = "minecraft.png";
    private static final String DESKTOP_FILE_NAME = APP_ID + ".desktop";
    private static final String DESKTOP_FILE_RESOURCE = "/assets/waygl/" + DESKTOP_FILE_NAME;

    private final List<Path> injected = new ArrayList<>();

    private IconInjector() {}

    public void inject(String minecraftVersion) {
        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
            uninjectFiles();

            updateIcons();
        }));

        try (var stream = IconInjector.class.getResourceAsStream(DESKTOP_FILE_RESOURCE)) {
            if (stream == null) {
                LOGGER.error("Icon resource not found: " + DESKTOP_FILE_RESOURCE);
                return;
            }

            var location = getDesktopFileLocation();

            var data = new String(stream.readAllBytes(), StandardCharsets.UTF_8);
            var formatted = String.format(data, minecraftVersion, "minecraft");

            injectFile(location, formatted.getBytes(StandardCharsets.UTF_8));
        } catch (IOException e) {
            LOGGER.error("Could not write icon file", e);
        }
    }

    public void setIcon(Collection<InputStream> icons) {
        try {
            for (var icon : icons) {
                if (icon == null) continue;

                var iconData = icon.readAllBytes();

                var image = ImageIO.read(new ByteArrayInputStream(iconData));
                var target = getIconFileLocation(image.getWidth(), image.getHeight());

                injectFile(target, iconData);
            }
        } catch (IOException e) {
            LOGGER.error("Failed to set icon: ", e);
            return;
        }

        updateIcons();
    }

    private void injectFile(Path target, byte[] data) {
        try {
            Files.createDirectories(target.getParent());
            Files.write(target, data);
            injected.add(target);
        } catch (IOException e) {
            LOGGER.error("Failed to create file: {}", target);
            LOGGER.error(e.toString());
        }
    }

    private void uninjectFiles() {
        for (var path : injected) {
            try {
                Files.deleteIfExists(path);
            } catch (IOException e) {
                LOGGER.error("Failed to delete file: {}", path, e);
            }
        }
    }

    public static IconInjector getInstance() {
        return INSTANCE;
    }

    private static Path getIconFileLocation(int width, int height) {
        return XDGUtils.getUserDataLocation()
                .resolve("icons/hicolor")
                .resolve(width + "x" + height)
                .resolve("apps")
                .resolve(ICON_FILE_NAME);
    }

    private static Path getDesktopFileLocation() {
        return XDGUtils.getUserDataLocation()
                .resolve("applications")
                .resolve(DESKTOP_FILE_NAME);
    }

    private static void updateIcons() {

        try {
            String[] xdgIconResourceCmdLine = Loader.isFlatpak() ?
                    new String[] { "flatpak-spawn", "--host", "xdg-icon-resource", "forceupdate" } : new String[] { "xdg-icon-resource", "forceupdate" };

            var xdgUpdateTask = new ProcessBuilder(xdgIconResourceCmdLine);
            xdgUpdateTask.start();
        } catch (IOException e) {
            LOGGER.warn("Failed to update icon theme", e);
        }
    }
}

File: mod/common/src/main/kotlin/net/wiredtomato/waygl/IconInjector.kt
package net.wiredtomato.waygl

import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.awt.image.BufferedImage
import java.io.InputStream
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.nio.file.Path
import javax.imageio.ImageIO

/**
 * Port of [moehreag/wayland_fixes](https://github.com/moehreag/wayland-fixes) DesktopFileInjector
 */
object IconInjector {
    const val APP_ID = "com.mojang.minecraft"
    private const val ICON_NAME = "minecraft.png"
    private const val FILE_NAME = "$APP_ID.desktop"
    private const val LOCATION = "/assets/waygl/$FILE_NAME"
    private val injects = mutableListOf<Path>()
    private val LOGGER: Logger = LoggerFactory.getLogger("WayGL/IconInjector")

    fun inject(minecraftVersion: String) {
        Runtime.getRuntime().addShutdownHook(Thread(IconInjector::uninject))

        val stream = IconInjector::class.java.getResourceAsStream(LOCATION)
        val location = getDesktopFileLoc()

        injectFile(
            location,
            String.format(stream?.readAllBytes()?.toString(Charsets.UTF_8) ?: "null",
                minecraftVersion,
                ICON_NAME.take(ICON_NAME.lastIndexOf("."))
            ).toByteArray(StandardCharsets.UTF_8)
        )
    }

    private fun uninject() {
        injects.forEach {
            { Files.deleteIfExists(it) }.runCatching {
                this()
            }
        }

        updateIconSys()
    }

    fun setIcon(icons: Collection<InputStream>) {
        val result = {
            icons.forEach {
                val image: BufferedImage = ImageIO.read(it)
                val target: Path = getIconFileLoc(
                    image.width,
                    image.height
                )
                injectFile(target, it.readAllBytes())
            }
        }.runCatching {
            this()
        }

        if (result.isFailure) {
            throw result.exceptionOrNull()!!
        }
    }

    private fun injectFile(target: Path, data: ByteArray) {
        val result = {
            Files.createDirectories(target.parent)
            Files.write(target, data)
            injects.add(target)
        }.runCatching {
            this()
        }

        if (result.isFailure) {
            LOGGER.error("Failed to inject file: $target")
            LOGGER.error(result.exceptionOrNull().toString())
        }
    }

    private fun getIconFileLoc(width: Int, height: Int) =
        XDG.getUserDataLocation()
            .resolve("icons/hicolor")
            .resolve("${width}x$height")
            .resolve("apps")
            .resolve(ICON_NAME)

    private fun getDesktopFileLoc() =
        XDG.getUserDataLocation()
            .resolve("applications")
            .resolve(FILE_NAME)

    private fun updateIconSys() {
        ProcessBuilder("xdg-icon-resource", "forceupdate").runCatching {
            start()
        }
    }
}